### PR TITLE
API: Added ExtensionArray constructors

### LINKED ISSUE #19906: ExtensionArray constructor issue
I don't think that pandas should make any assumptions on how subclasses `__init__` method. Currently we assume that the first positional argument accepts an instance of the class or a sequence of the class's scalar type.

I'd rather break that into two distinct methods.

```python
@classmedthod
def from_extension_array(cls, extension_array: ExtensionArray) -> ExtensionArray:
    """Construct a new ExtensionArray from an instance of the same type

    Parameters
    ----------
    extension_array : ExtensionArray
        An instance of 'cls'

    Returns
    -------
    ExtensionArray
    """


ScalarType = ExtensionArray.dtype.type

@classmethod
def from_scalars(cls, scalars: Sequence[ScalarType]) -> ExtensionArray:
    """Construct a new ExtensionArray from a sequence of the scalar type

    Parameters
    ----------
    scalars : Sequence[ScalarType]
        A sequence of cls.dtype.type, the scalar type for this array

    Returns
    -------
    ExtensionArray
    """
```

I think these should be abstract. For many subclasses, a simple `cls(arg)` should suffice.

### PR DESCRIPTION
**Pull request by TomAugspurger**
Adds two new (private, but part of the interface) constructors to EA.

Closes #19906

### CODE REVIEW

**Comment by jreback** [comment]
this is just API bloat. what is the point here?

**Review by jreback — CHANGES_REQUESTED** [review]
comments

**Comment by TomAugspurger** [comment]
> what is the point here?

Not imposing things on the subclasses `__init__`. I believe @jorisvandenbossche already had issues with the old design in geopandas. You could imagine a class whose `__init__` wants to use keyword only arguments.

**Comment by jorisvandenbossche** [comment]
@jreback I think this is actually very much in line with how we are doing it currently internally (having constructors as class methods). The only difference here is that we make a constructor class method that is more specific in its behaviour, which I think is a plus both for internal usage as for the extension author to clearly communicate what we exactly expect from the constructor (in this case, that it takes a sequence of scalars). 
I think limiting the expected behaviour is a good thing. And we can call it `_constructor` to be 'consistent' with other places in the code base, but given its specific behaviour, I think embedding that in the name is better (we could also use `_constructor_from_scalars` if you like that more).

**Comment by jorisvandenbossche** [comment]
> Not imposing things on the subclasses __init__. I believe @jorisvandenbossche already had issues with the old design in geopandas. You could imagine a class whose __init__ wants to use keyword only arguments.

To add: it is not *really* a problem for geopandas to support this in its `__init__.py`. It was just that we initially did not choose to do this, and I needed to adapt the `__init__;py` in my branch that tests the extension array interface (and for that I just needed to figure out some memory management stuff that Matthew originally did). But now it is coded, it is also not complicated to support it. 
But that said, I still think it is useful for the extension interface to be very explicit about what we expect from certain constructors, so having a separate one for constructing from scalars is IMO a good idea (giving more flexibility about the main `__init__` to the extension author).

**Comment by TomAugspurger** [comment]
I think `_from` already indicates pretty well that it's a constructor. `_constructor_from` feels redundant.

**Comment by TomAugspurger** [comment]
> this is just API bloat. what is the point here?

@jreback does the comments https://github.com/pandas-dev/pandas/pull/19913#issuecomment-368785002 and https://github.com/pandas-dev/pandas/pull/19913#issuecomment-368728194 make sense? This is blocking https://github.com/pandas-dev/pandas/pull/19869 and factorizing EAs.

**Comment by jreback** [comment]
> @jreback I think this is actually very much in line with how we are doing it currently internally (having constructors as class methods). The only difference here is that we make a constructor class method that is more specific in its behaviour, which I think is a plus both for internal usage as for the extension author to clearly communicate what we exactly expect from the constructor (in this case, that it takes

I am not opposing this motivation. I agree with it. But you going off with new names w/o a very clear purpose. We *already* have somewhat of a mess, e.g. ``_shallow_copy`` and ``_simple_new`` in the Index space, and ``_constructor`` in the Series/DataFrame space. So you are creating something new here. Let's stick with the current ones. Then if you want to change they should be changed for everything. Forging yet another path is just going creating technical debt.

This is my basic hesitiion with things like creating a new ``.fillna`` impl as well. Sure adding code is exciting as it can be very clean. But the existing code base is simply being let behind, creating an *even bigger* issue for later.

**Comment by TomAugspurger** [comment]
> But you going off with new names w/o a very clear purpose

Sorry I wasn't clearer then. But I thought the convention was that `self._constructor` was a property that returns a class. The proposal here is for a classmethod. We have places in the code that do things expect `_constructor` to be a class like `self._constructor.__name__`.

I don't really care whether we call it `_constructor` or `_from_scalars`. But they're different things so I'd prefer to have different names. Your call.

**Comment by jorisvandenbossche** [comment]
@jreback ping

**Comment by jreback** [comment]
I dont' see a purpose of adding this, nor are you showing utility.

**Comment by jorisvandenbossche** [comment]
I think all our long comments are trying to explain the utility. And remember it was you who asked to add a constructor class method in the first place. The last question of Tom (the "your call") was mainly about the *name*. 
Unless you disagree about adding a constructor like this that is limited in functionality (a specific one to construct from scalars) at all. But then please be more specific.

**Comment by TomAugspurger** [comment]
> simply need `_constructor` as a class property

Ah, I see the misunderstanding. That's not an option since it relies on `self._constructor(data)` being valid. We don't know that the subclass's `__init__` takes a sequence of scalar types.

**Comment by TomAugspurger** [comment]
Trying to summarize the arguments here.

Jeff wants a property `_constructor` that constructs the new EA. This would have the same requirements used in pandas today

1. `self._constructor` is a class
2. The class can be constructed with a single positional argument.

My objection to that is that I don't want to put any restrictions a 3rd party EAs `__init__` method. For example, an IPArray might be constructable in a few ways

1. sequence of `ipaddress` objects
2. sequence of integers
3. a bytestring

I want to support all these, but not with pandas' "magical" constructors of just inferring things. So I use keyword only arguments.

```python
class IPArray(ExtensionArray):
    """Holder for IP Addresses."""
    def __init__(self, *, addresses=None, ints=None, bytestring=None):
        from .parser import _to_ip_array

        if addresses is not None:
            values = _to_ip_array(addresses)
        elif ints is not None:
            ...
        else:
            values = self._from_bytes(addresses)
        self.data = values
```

We can't use pandas usual `_constructor` route

```python
@property
def _constructor(self):
    return IPArray
```

What's *possible* is to use a property that returns a function

```python
    @property
    def _constructor(self):
        def fn(values):
            return type(self)(addresses=values)
        return fn
```

Now, the question is, do we *want* this? This now puts restrictions on uses of `self._constructor` within pandas.

1. We need to ensure we're just using `self._constructor` as a callable. We can't treat it as a class to access additional attributes off it.
2. We're conflating two similar things:
  a.) A method for creating a data structure from a very specific type (sequence of your scalar type)
  b.) A method for creating a data structure from anything.

Again, I don't feel too strongly about this, but I do prefer `_from_scalars` since it's clearer that this is a very specific use case that will be limited to just `arrays/base.py`.

**Comment by jreback** [comment]
I think the problem with this is the name. I understand and am ok with a class method that allows an extension array author to do internal construction with a class method, eg.. ``_constructor_from_default(....)``, The ``_from_scalars`` is just confusing, because you are most often costructing from an ``ndarray`` or a ``list`` (e.g. after an internal pandas operation) of *scalars*.

**Comment by jorisvandenbossche** [comment]
I don't really mind the exact name, but would like to keep the word 'scalar' in it, as that is essential (it is not just accepting an array/list, but an array/list of scalars). 
The longest and complete name would be `_constructor_from_sequence_of_scalars` ... so but now we need to find the best compromise between length and clarity :-)

**Comment by jreback** [comment]
and I would like to remove scalars, yes its *from* a sequence of scalars, the relevant part is that its a list of *something*.

``_constructor_from_sequence`` works for me.

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ class ExtensionArray(object):
     The interface includes the following abstract methods that must be
     implemented by subclasses:
 
+    * _from_scalars
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py]: update this one as well

**Comment by TomAugspurger** [comment]
All green. merging later today.

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**